### ISSUE TITLE
Weird production crash

### ISSUE THREAD

**Issue by omerlh** [jkolb/FieryCrucible#16]
Hey
Recently we noticed a few production crash, that we are unable to reproduce. This is part of the callstack (the top of the stack, only the FieryCrucible code):
````
0
libswiftCore.dylib	
swift_unknownRelease + 8
1	***	
_TTSf4n_n_n_n_d___TTSg5GVs12_ArrayBufferVC13***17DependencyFactoryP33_35ADD3F8F67F80346DAF89801BD14AB811InstanceKey_GS_S2__s20_ArrayBufferProtocols_GVs14_IgnorePointerS2__GS4_S2__s16_PointerFunctions___TFs22_arrayOutOfPlaceUpdateu0_Rxs20_ArrayBufferProtocol_s16_PointerFunctionwx5IndexzSiwx7Elementzw_7ElementrFTRxRGVs22_ContiguousArrayBufferwxS2__SiSiq__T_ + 276
2	***	
_TTSg5FT_T____TFSa16_copyToNewBufferfT8oldCountSi_T__merged + 84
3	***	
_TFC13***17DependencyFactoryP33_35ADD3F8F67F80346DAF89801BD14AB86injectu0_R_S_P33_35ADD3F8F67F80346DAF89801BD14AB817InstanceContainerxzw_12InstanceTyperfT8lifecyleOS0_P33_35ADD3F8F67F80346DAF89801BD14AB88Lifecyle4nameSS12instancePoolRGVs10DictionarySSPs9AnyObject__16containerFactoryFxq_7factoryKT_x9configureGSqFxT___x + 2348
4	***	
_TFC13***17DependencyFactory8unsharedurfTSS7factoryKT_x9configureGSqFxT___x + 404
5	***	
_TFC13***17DependencyFactory6sharedurfTKT_x4nameSS9configureGSqFxT___x_merged + 200
6
***	
_TFC13***17DependencyFactory8unsharedurfTKT_x4nameSS9configureGSqFxT___x + 28
````
I replaced our project name with ***.
And the reason to crash is `EXC_BAD_ACCESS KERN_INVALID_ADDRESS 0x0000000000000002`.
Also, we did not notice specific OS version or specific model. 
Could you help us shed some light on that crash?

Thanks,
Omer

**Comment by jkolb** [comment]
Unshared instances during the request are stored inside of a temporary dictionary and then that dictionary is discarded at the end of the request. My guess is there may be a chance where the temporary dictionary is released earlier than expected that may trigger the issue. It may happen more often in optimized production code. There isn't much to find when searching for swift_unknownRelease on the Internet.

Weak or unowned references being used to store returned instances may also cause this issue as the compiler may decided to release the value before the rest of the program is finished using it.

Also if there are multiple threads involved that could definitely be related as FieryCrucible isn't thread safe on its own. To make it so I would suggest adding locking inside any factory methods that are being used from multiple threads, or to have one factory per thread (shared instances would then be thread specific).

Let me see if I can come up with a test case that might trigger the issue as that will definitely help with fixing it. It looks like you are using an older version of FieryCrucible, not sure that would make a difference, but it would make fixing the issue easier for me if I only have to update the latest version.

**Comment by omerlh** [comment]
Hey,
First, thank you for the detailed response!
I know we are using an outdated version, that why I asked about cocoapod support. Now we are using the latest pod version, and it should be released in the next weeks. 
We also suspected something regarding thread safety - the crash happens on a different thread. I also tried to move the code that use the factory back to the main thread, it should also be released in the next weeks.
I will update after we release this version, as it seems that we both suspect the same issue - thread safety - so I hope it will solve it.
